### ISSUE TITLE
Docker daemon crashes when "docker exec -d" spawns background processes

### ISSUE THREAD

**Issue by jaylinski** [docker/docker#26785]
**Description**

This bug has already been reported here: https://github.com/docker/docker/issues/25989

But the solution provided in the previous ticket (use the detached mode) did not work for me.

The docker daemon still crashes.

**Steps to reproduce the issue:**
1. Run a script which spawns another script: `docker-compose exec -d -T <container_name> ./script &` (see below)
2. Try to stop container `docker-compose stop`.
3. Docker daemon is not responding.

These are docker-compose commands, but since docker-compose is using standard docker commands in background, this should also apply to non-compose commands.

**script**
```
#!/bin/bash
sleep 1000000 &
sh ./child
```
**child**
```
#!/bin/bash
sleep 1000000 &
```

**Describe the results you received:**
Docker daemon crashes.

**Describe the results you expected:**
Docker daemon does not crash and script keeps running in background.

**Output of `docker version`:**

```
Docker version 1.12.1, build 23cf638
```

**Output of `docker-compose version`:**

```
docker-compose version 1.8.0, build f3628c7
```

**Output of `docker info`:**

```
Containers: 6
 Running: 4
 Paused: 0
 Stopped: 2
Images: 34
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-253:1-1177880-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 3.242 GB
 Data Space Total: 107.4 GB
 Data Space Available: 34.04 GB
 Metadata Space Used: 6.271 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.141 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.110 (2015-10-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-36-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953 GiB
Name: xxx
ID: HPT2:FMT2:G375:LIMN:WZME:P2CL:2LSX:Z7DC:RCQF:2FX5:AMAI:6Q6Z
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Tested on Ubuntu 16.04 DigitalOcean Droplet.

**Comment by thaJeztah** [comment]
Ping @dnephin @mlaventure

**Comment by dnephin** [comment]
> Docker daemon crashes.

Please include the stack trace from from the log.

**Comment by jaylinski** [comment]
@dnephin when running `sudo journalctl -fu docker.service` I see no stacktrace.

So the docker daemon apperently does not crash, but is becoming unresponsive.

Just for completeness, I recreated the issue without docker-compose:

1. `docker run -d --name testcontainer php:7-fpm`
2. `docker exec -it testcontainer /bin/bash -c "TERM=$TERM exec bash"`   
Create the `script` and `child` files as described above and make them executable.
3. `docker exec -d testcontainer ./script`
4. `docker stop testcontainer`
5. No response from daemon.

This is what I get when restarting the service with `sudo systemctl restart docker`:

```
Sep 21 22:14:19 ubuntu dockerd[11781]: time="2016-09-21T22:14:19.577498710+02:00" level=error msg="Force shutdown daemon"
Sep 21 22:14:19 ubuntu dockerd[11781]: time="2016-09-21T22:14:19.577686029+02:00" level=info msg="Failed to send signal 15 to the process, force killing"
Sep 21 22:14:19 ubuntu dockerd[11781]: time="2016-09-21T22:14:19.577742964+02:00" level=error msg="Handler for POST /v1.24/containers/testcontainer/stop returned error: Cannot stop container testcontainer: Cannot kill container 130e1237ae33a2724de2d9ef10c1325fcfe4b81e80447e335583cfca4dc2347e: rpc error: code = 2 desc = grpc: the client connection is closing"
Sep 21 22:14:19 ubuntu dockerd[11781]: time="2016-09-21T22:14:19.57774751+02:00" level=info msg="stopping containerd after receiving terminated"
Sep 21 22:14:19 ubuntu dockerd[11781]: time="2016-09-21T22:14:19.577969466+02:00" level=info msg="Failed to send signal 15 to the process, force killing"
Sep 21 22:14:19 ubuntu dockerd[11781]: time="2016-09-21T22:14:19.578004744+02:00" level=error msg="Stop container error: Failed to stop container 130e1237ae33a2724de2d9ef10c1325fcfe4b81e80447e335583cfca4dc2347e with error: Cannot kill container 130e1237ae33a2724de2d9ef10c1325fcfe4b81e80447e335583cfca4dc2347e: rpc error: code = 2 desc = grpc: the client connection is closing"
Sep 21 22:14:20 ubuntu systemd[1]: Stopped Docker Application Container Engine.
Sep 21 22:14:20 ubuntu systemd[1]: Starting Docker Application Container Engine...
Sep 21 22:14:20 ubuntu dockerd[12691]: time="2016-09-21T22:14:20.617112804+02:00" level=info msg="libcontainerd: new containerd process, pid: 12698"
Sep 21 22:14:21 ubuntu dockerd[12691]: time="2016-09-21T22:14:21.676694812+02:00" level=info msg="[graphdriver] using prior storage driver \"aufs\""
Sep 21 22:14:21 ubuntu dockerd[12691]: time="2016-09-21T22:14:21.719882177+02:00" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Sep 21 22:14:21 ubuntu dockerd[12691]: time="2016-09-21T22:14:21.720095385+02:00" level=warning msg="Your kernel does not support swap memory limit."
Sep 21 22:14:21 ubuntu dockerd[12691]: time="2016-09-21T22:14:21.720486539+02:00" level=info msg="Loading containers: start."
Sep 21 22:14:21 ubuntu dockerd[12691]: ........................time="2016-09-21T22:14:21.736287902+02:00" level=warning msg="libcontainerd: client is out of sync, restore was called on a fully synced container (130e1237ae33a2724de2d9ef10c1325fcfe4b81e80447e335583cfca4dc2347e)."
Sep 21 22:14:21 ubuntu dockerd[12691]: time="2016-09-21T22:14:21.829725491+02:00" level=info msg="Firewalld running: false"
Sep 21 22:14:22 ubuntu dockerd[12691]: time="2016-09-21T22:14:22.211598928+02:00" level=info msg="Removing stale sandbox f74e5ec0942033b657e3e6179380cd50d83518a0fda4f4d82018939e37b840b4 (130e1237ae33a2724de2d9ef10c1325fcfe4b81e80447e335583cfca4dc2347e)"
Sep 21 22:14:22 ubuntu dockerd[12691]: time="2016-09-21T22:14:22.291714685+02:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Sep 21 22:14:22 ubuntu dockerd[12691]: time="2016-09-21T22:14:22.352527217+02:00" level=info msg="Loading containers: done."
Sep 21 22:14:22 ubuntu dockerd[12691]: time="2016-09-21T22:14:22.352593584+02:00" level=info msg="Daemon has completed initialization"
Sep 21 22:14:22 ubuntu dockerd[12691]: time="2016-09-21T22:14:22.352608702+02:00" level=info msg="Docker daemon" commit=23cf638 graphdriver=aufs version=1.12.1
Sep 21 22:14:22 ubuntu systemd[1]: Started Docker Application Container Engine.
Sep 21 22:14:22 ubuntu dockerd[12691]: time="2016-09-21T22:14:22.358094266+02:00" level=info msg="API listen on /var/run/docker.sock"
```

**Comment by mlaventure** [comment]
This is the issue caused by exec's children not being reaped correctly. 

This was fixed when #26538 was merged.

It now works fine on master for me.
